### ISSUE TITLE
I'm having some kind of key trust issue running ./ffbuild.sh

### ISSUE THREAD

**Issue by JaiSoone** [jtanx/fontforgebuilds#9]
The following sequence would seem to indicate that it wanted me to answer yes or no,
but it didn't wait for a reply and then got into some kind of trouble. Do I have to tweak my firewall?
Here is the console listing:
$ ./ffbuild.sh -d
MSYS2 FontForge build script...
Building 64-bit version!
First time run; installing MSYS and MinGW libraries...
Adding the fontforgelibs repo...
gpg: requesting key 2E618C78 from hkp server pool.sks-keyservers.net
gpg: key 2E618C78: public key "Jeremy Tan <jtanx@outlook.com>" imported
gpg: Total number processed: 1
gpg:               imported: 1  (RSA: 1)
==> Updating trust database...
gpg: no need for a trustdb check
  -> Locally signing key 2E618C78...
==> Updating trust database...
gpg: 3 marginal(s) needed, 1 complete(s) needed, PGP trust model
gpg: depth: 0  valid:   1  signed:   5  trust: 0-, 0q, 0n, 0m, 0f, 1u
gpg: depth: 1  valid:   5  signed:   3  trust: 1-, 0q, 0n, 4m, 0f, 0u
gpg: depth: 2  valid:   3  signed:   0  trust: 3-, 0q, 0n, 0m, 0f, 0u
:: Synchronizing package databases...
 mingw32 is up to date
 mingw64 is up to date
 msys is up to date
 fontforgelibs32                     14.5 KiB  0.00B/s 00:00 [###############################] 100%
 fontforgelibs32.sig                543.0   B  0.00B/s 00:00 [###############################] 100%
error: fontforgelibs32: key "25D7E9C490F90C4A" is unknown
:: Import PGP key 4096R/E10ABC00BD686519D6AC5F5D25D7E9C490F90C4A, "Jeremy Tan <jtanx@outlook.com>", created: 2016-08-27? [Y/n]
error: fontforgelibs32: signature from "Jeremy Tan <jtanx@outlook.com>" is unknown trust
error: failed to update fontforgelibs32 (invalid or corrupted database (PGP signature))
 fontforgelibs64                     14.5 KiB  0.00B/s 00:00 [###############################] 100%
 fontforgelibs64.sig                543.0   B  0.00B/s 00:00 [###############################] 100%
error: fontforgelibs64: signature from "Jeremy Tan <jtanx@outlook.com>" is unknown trust
error: failed to update fontforgelibs64 (invalid or corrupted database (PGP signature))

**Comment by jtanx** [comment]
It looks like you have a really old version of ffbuild.sh.  I changed PGP keys from 2E618C78 to 90F90C4A back in 2016. Have you done a git pull?

You'll then have to run:

```
pacman-key -r 90F90C4A
```

To add the new key.

Your /etc/pacman.conf should also have this at the end:

```
[fontforgelibs32]
Server = https://dl.bintray.com/jtanx/fontforgelibs/fontforgelibs32
Server = http://downloads.sourceforge.net/project/fontforgebuilds/build-system-extras/fontforgelibs/i686
[fontforgelibs64]
Server = https://dl.bintray.com/jtanx/fontforgelibs/fontforgelibs64
Server = http://downloads.sourceforge.net/project/fontforgebuilds/build-system-extras/fontforgelibs/x86_64
```

(Note the bintray server takes precedence over sourceforge; it's generally more reliable to download from)

**Comment by JaiSoone** [comment]
Well, I'm going to start over again for the third time. I just looked at what I have in ff-build/ReleasePackage/CHANGELOG.txt and it's dated beginning with 04-07-15 (14-06-15) (r4):
which would seem to be quite old comparing it with what I see in:
https://github.com/fontforge/fontforgebuilds/blob/master/ReleasePackage/CHANGELOG.txt
namely, a date beginning with 17/06/17:. I like playing with a full deck.

To answer you question, I would guess that I got an old version of ffbuild.sh from Build procedure 
5 )  Once installed, you can now clone the build script:
      ~~~~~~
       git clone git://git.code.sf.net/p/fontforgebuilds/code ff-build

based on the comments, but I don't know that for sure. I did do optional Step 6 without having Notepad++ installed first. I was guessing that not having Notepad++ wouldn't actually make a difference. I do have Notepad++ installed though at this point. I'm going to stop after step 5 and see what I've ended up with before proceeding. I come from an HP-UX background; so, I'm vaguely familiar with what's happening here. I have a large enough disc drive; so, I could dual boot it and put a Linux distribution of some sort in one of the partitions, but I'm not particularly interested in figuring out how to make that all work. I'll keep you posted. Thanks for the response.

**Comment by jtanx** [comment]
Oh yeah, that sourceforge git repository isn't up to date, sorry about that. You probably want to clone this one instead:

```
git clone https://github.com/jtanx/fontforgebuilds ff-build
```

edit: I manually synced the Sourceforge repo, but I would recommend to use the above listed command (e.g. clone this github repository).

**Comment by JaiSoone** [comment]
I've gotten a lot further along now. I'm now blowing up on:
Finished installing precompiled libraries!
Installing Cairo...
Installing cairo-1.15.6...
tar: /home/NateJoe/ff-build/original-archives/sources/cairo-1.15.6.tar.xz: Cannot open: No such file or directory
tar: Error is not recoverable: exiting now
!!! Build failed at: cairo-1.15.6

So, I unzipped and extracted the source files myself. It seems that the script wants to do that itself, right?
Is it going to want to do that for the binaries as well? Also, is this script hard coded to want pango 1.40.6?
pango 1.40.7 is now available.

**Comment by jtanx** [comment]
Yes, all you should need to do is to place the archive files in the right folder, e.g. /home/NateJoe/ff-build/original-archives/sources/cairo-1.15.6.tar.xz and ffbuild.sh will handle extracting them.

Yes, the versions used are hardcoded into ffbuild.sh, but if you want to use a later version, it's easy enough to find all occurrences of 1.40.6 and replace them with 1.40.7 in ffbuild.sh.

Note however, that you can skip compiling Pango and Cairo altogether and use a precompiled package. This will make ffbuild.sh skip straight to compiling FontForge itself. To do this, use:

```
./ffbuild.sh --precompiled-pango-cairo 
```

You can see other options by running

```
./ffbuild.sh --help
```

**Comment by JaiSoone** [comment]
OK. I put the archive files in place and started building again. It now runs into trouble trying to compile some part of cairo. I now end up with:

  CC       cairo-xcb-shm.lo
  CC       cairo-xcb-surface.lo
  CC       cairo-xcb-surface-core.lo
In file included from C:/msys64/mingw64/include/xcb/xcb.h:34:0,
                 from cairo-xcb.h:46,
                 from cairo-xcb-private.h:42,
                 from cairo-xcb-connection.c:35:
C:/msys64/mingw64/include/xcb/xcb_windefs.h:36:10: fatal error: X11/XTrans/Xtrans.h: No such file or directory
 #include <X11/XTrans/Xtrans.h>
          ^~~~~~~~~~~~~~~~~~~~~
compilation terminated.
In file included from C:/msys64/mingw64/include/xcb/xcb.h:34:0,
                 from cairo-xcb.h:46,
                 from cairo-xlib-xcb-surface.c:44:
C:/msys64/mingw64/include/xcb/xcb_windefs.h:36:10: fatal error: X11/XTrans/Xtrans.h: No such file or directory
 #include <X11/XTrans/Xtrans.h>
          ^~~~~~~~~~~~~~~~~~~~~
compilation terminated.
make[3]: *** [Makefile:2348: cairo-xcb-connection.lo] Error 1
make[3]: *** Waiting for unfinished jobs....
make[3]: *** [Makefile:2348: cairo-xlib-xcb-surface.lo] Error 1
In file included from C:/msys64/mingw64/include/xcb/xcb.h:34:0,
                 from cairo-xcb.h:46,
                 from cairo-xcb-private.h:42,
                 from cairo-xcb-connection-core.c:34:
C:/msys64/mingw64/include/xcb/xcb_windefs.h:36:10: fatal error: X11/XTrans/Xtrans.h: No such file or directory
 #include <X11/XTrans/Xtrans.h>
          ^~~~~~~~~~~~~~~~~~~~~
compilation terminated.
make[3]: *** [Makefile:2349: cairo-xcb-connection-core.lo] Error 1
In file included from C:/msys64/mingw64/include/xcb/xcb.h:34:0,
                 from cairo-xcb.h:46,
                 from cairo-xcb-private.h:42,
                 from cairo-xcb-connection-render.c:34:
C:/msys64/mingw64/include/xcb/xcb_windefs.h:36:10: fatal error: X11/XTrans/Xtrans.h: No such file or directory
 #include <X11/XTrans/Xtrans.h>
          ^~~~~~~~~~~~~~~~~~~~~
compilation terminated.
make[3]: *** [Makefile:2348: cairo-xcb-connection-render.lo] Error 1
In file included from C:/msys64/mingw64/include/xcb/xcb.h:34:0,
                 from cairo-xcb.h:46,
                 from cairo-xcb-private.h:42,
                 from cairo-xcb-screen.c:35:
C:/msys64/mingw64/include/xcb/xcb_windefs.h:36:10: fatal error: X11/XTrans/Xtrans.h: No such file or directory
 #include <X11/XTrans/Xtrans.h>
          ^~~~~~~~~~~~~~~~~~~~~
compilation terminated.
make[3]: *** [Makefile:2348: cairo-xcb-screen.lo] Error 1
In file included from C:/msys64/mingw64/include/xcb/xcb.h:34:0,
                 from cairo-xcb.h:46,
                 from cairo-xcb-surface.c:43:
C:/msys64/mingw64/include/xcb/xcb_windefs.h:36:10: fatal error: X11/XTrans/Xtrans.h: No such file or directory
 #include <X11/XTrans/Xtrans.h>
          ^~~~~~~~~~~~~~~~~~~~~
compilation terminated.
make[3]: *** [Makefile:2348: cairo-xcb-surface.lo] Error 1
In file included from C:/msys64/mingw64/include/xcb/xcb.h:34:0,
                 from cairo-xcb.h:46,
                 from cairo-xcb-private.h:42,
                 from cairo-xcb-surface-core.c:35:
C:/msys64/mingw64/include/xcb/xcb_windefs.h:36:10: fatal error: X11/XTrans/Xtrans.h: No such file or directory
 #include <X11/XTrans/Xtrans.h>
          ^~~~~~~~~~~~~~~~~~~~~
compilation terminated.
make[3]: *** [Makefile:2348: cairo-xcb-surface-core.lo] Error 1
make[3]: Leaving directory '/home/NateJoe/ff-build/work/mingw64/cairo-1.15.6/src'
make[2]: *** [Makefile:1947: all] Error 2
make[2]: Leaving directory '/home/NateJoe/ff-build/work/mingw64/cairo-1.15.6/src'
make[1]: *** [Makefile:904: all-recursive] Error 1
make[1]: Leaving directory '/home/NateJoe/ff-build/work/mingw64/cairo-1.15.6'
make: *** [Makefile:755: all] Error 2
!!! Build failed at: cairo-1.15.6

NateJoe@TinkeringAround MINGW64 ~/ff-build
$


This should work, right?
I think I'll switch over to using the precompiled versions.
If I switch to using the build option that calls for using the precompiled versions,
can I leave the source archives in place anyway, or do I have to remove them?

**Comment by jtanx** [comment]
That would suggest that the X11 packages haven't been installed. I think I know what's going on - because you ran ffbuild.sh earlier, but the older version didn't have the precompiled packages, it hasn't been installed yet.

Under the same folder as ffbuild.sh, it has made a file, likely called `.pacman-mingw32-installed` or `.pacman-mingw64-installed` - remove both if present. This will cause ffbuild.sh to reinstall any dependencies that it needs. 

Now that I recall, it might be best to not install the precompiled Pango/Cairo, because this will affect your msys2 environment globally, and if you build any other packages under this same environment, where they depend on Pango/Cairo, it introduces an unnecessary dependency on X11 components, which is only really needed by FontForge.

In any case, it's fine to leave archive files in place even if using precompiled versions. So to summarise, I recommend doing something like:

```
rm -f .pacman-mingw32-installed
rm -f .pacman-mingw64-installed
./ffbuild.sh
```

**Comment by JaiSoone** [comment]
It seems to be a little bit more complicated than that. I did have a .pacman-mingw64-installed file but not the other because I've only been trying 64 bit builds. I removed it and ran ffbuild.sh again, but it still fails. Here is the complete console listing (it fails rather quickly thank goodness):
$ ./ffbuild.sh -d
MSYS2 FontForge build script...
Building without the GDK backend (using X11 instead).
Building 64-bit version!
First time run; installing MSYS and MinGW libraries...
:: Synchronizing package databases...
 mingw32 is up to date
 mingw64 is up to date
 msys is up to date
 fontforgelibs32 is up to date0   B  0.00B/s 00:00 [---------------------]   0%
 fontforgelibs64 is up to date0   B  0.00B/s 00:00 [---------------------]   0%
warning: diffutils-3.5-1 is up to date -- skipping
warning: findutils-4.6.0-1 is up to date -- skipping
warning: make-4.2.1-1 is up to date -- skipping
warning: patch-2.7.5-1 is up to date -- skipping
warning: tar-1.29-1 is up to date -- skipping
warning: automake-wrapper-10-1 is up to date -- skipping
warning: autoconf-2.69-3 is up to date -- skipping
warning: pkg-config-0.29.2-1 is up to date -- skipping
 there is nothing to do
warning: mingw-w64-x86_64-gcc-7.1.0-2 is up to date -- skipping
warning: mingw-w64-x86_64-gmp-6.1.2-1 is up to date -- skipping
warning: mingw-w64-x86_64-ntldd-git-r15.e7622f6-2 is up to date -- skipping
warning: mingw-w64-x86_64-gettext-0.19.8.1-2 is up to date -- skipping
warning: mingw-w64-x86_64-libiconv-1.15-1 is up to date -- skipping
warning: mingw-w64-x86_64-libtool-2.4.6-9 is up to date -- skipping
 there is nothing to do
warning: mingw-w64-x86_64-python3-3.6.2rc1-2 is up to date -- skipping
warning: mingw-w64-x86_64-openssl-1.0.2.l-1 is up to date -- skipping
 there is nothing to do
Installing vanilla Pango and Cairo libraries...
warning: mingw-w64-x86_64-cairo-1.15.6-1 is up to date -- skipping
warning: mingw-w64-x86_64-pango-1.40.6-1 is up to date -- skipping
 there is nothing to do
Installing precompiled X11...
warning: mingw-w64-x86_64-libx11-git-r1325.7d2010fe-1 is up to date -- skipping
warning: mingw-w64-x86_64-libxext-git-r205.a07b4bb-1 is up to date -- skipping
 there is nothing to do
warning: mingw-w64-x86_64-libxrender-git-r82.f4f32f1-1 is up to date -- skipping
warning: mingw-w64-x86_64-libxft-git-r102.c418dc7-1 is up to date -- skipping
 there is nothing to do
Installing precompiled libspiro and libuninameslist...
warning: mingw-w64-x86_64-libspiro-git-r101.8cd70b2-1 is up to date -- skipping
warning: mingw-w64-x86_64-libuninameslist-git-r77.fd4e842-1 is up to date -- skipping
 there is nothing to do
Installing precompiled devel libraries...
warning: mingw-w64-x86_64-zlib-1.2.11-1 is up to date -- skipping
warning: mingw-w64-x86_64-libpng-1.6.30-1 is up to date -- skipping
warning: mingw-w64-x86_64-giflib-5.1.4-1 is up to date -- skipping
warning: mingw-w64-x86_64-libtiff-4.0.8-1 is up to date -- skipping
warning: mingw-w64-x86_64-libjpeg-turbo-1.5.1-1 is up to date -- skipping
warning: mingw-w64-x86_64-libxml2-2.9.4-4 is up to date -- skipping
 there is nothing to do
warning: mingw-w64-x86_64-freetype-2.8-1 is up to date -- skipping
warning: mingw-w64-x86_64-fontconfig-2.12.4-1 is up to date -- skipping
warning: mingw-w64-x86_64-glib2-2.52.3-1 is up to date -- skipping
warning: mingw-w64-x86_64-pixman-0.34.0-3 is up to date -- skipping
warning: mingw-w64-x86_64-harfbuzz-1.4.6-2 is up to date -- skipping
 there is nothing to do
Finished installing precompiled libraries!
Installing Cairo...
Installing cairo-1.15.6...
Sensing incomplete build, re-attempting the build...
Patching cairo-1.15.6 with cairo.patch...
Reversed (or previously applied) patch detected!  Skipping patch.
1 out of 1 hunk ignored
Reversed (or previously applied) patch detected!  Skipping patch.
1 out of 1 hunk ignored
Reversed (or previously applied) patch detected!  Skipping patch.
1 out of 1 hunk ignored
Reversed (or previously applied) patch detected!  Skipping patch.
1 out of 1 hunk ignored
Reversed (or previously applied) patch detected!  Skipping patch.
1 out of 1 hunk ignored
Reversed (or previously applied) patch detected!  Skipping patch.
1 out of 1 hunk ignored
Sensed that patch has already been applied; skipping
Sensed that the configure script has already run; delete cairo-1.15.6.configure-complete to rerun configure
 make -j9  || bail 'cairo-1.15.6'
Makefile:1496: warning: overriding recipe for target 'recheck'
Makefile:1134: warning: ignoring old recipe for target 'recheck'
make  all-recursive
make[1]: Entering directory '/home/NateJoe/ff-build/work/mingw64/cairo-1.15.6'
Makefile:1496: warning: overriding recipe for target 'recheck'
Makefile:1134: warning: ignoring old recipe for target 'recheck'
Making all in src
make[2]: Entering directory '/home/NateJoe/ff-build/work/mingw64/cairo-1.15.6/src'
make  all-am
make[3]: Entering directory '/home/NateJoe/ff-build/work/mingw64/cairo-1.15.6/src'
  CC       cairo-xlib-xcb-surface.lo
  CC       cairo-xcb-connection.lo
  CC       cairo-xcb-connection-core.lo
  CC       cairo-xcb-connection-render.lo
  CC       cairo-xcb-screen.lo
  CC       cairo-xcb-surface.lo
  CC       cairo-xcb-surface-core.lo
  CC       cairo-xcb-surface-render.lo
  CC       cairo-xcb-resources.lo
In file included from C:/msys64/mingw64/include/xcb/xcb.h:34:0,
                 from cairo-xcb.h:46,
                 from cairo-xcb-private.h:42,
                 from cairo-xcb-screen.c:35:
C:/msys64/mingw64/include/xcb/xcb_windefs.h:36:10: fatal error: X11/XTrans/Xtrans.h: No such file or directory
 #include <X11/XTrans/Xtrans.h>
          ^~~~~~~~~~~~~~~~~~~~~
compilation terminated.
In file included from C:/msys64/mingw64/include/xcb/xcb.h:34:0,
                 from cairo-xcb.h:46,
                 from cairo-xlib-xcb-surface.c:44:
C:/msys64/mingw64/include/xcb/xcb_windefs.h:36:10: fatal error: X11/XTrans/Xtrans.h: No such file or directory
 #include <X11/XTrans/Xtrans.h>
          ^~~~~~~~~~~~~~~~~~~~~
compilation terminated.
In file included from C:/msys64/mingw64/include/xcb/xcb.h:34:0,
                 from cairo-xcb.h:46,
                 from cairo-xcb-private.h:42,
                 from cairo-xcb-connection-core.c:34:
C:/msys64/mingw64/include/xcb/xcb_windefs.h:36:10: fatal error: X11/XTrans/Xtrans.h: No such file or directory
 #include <X11/XTrans/Xtrans.h>
          ^~~~~~~~~~~~~~~~~~~~~
compilation terminated.
In file included from C:/msys64/mingw64/include/xcb/xcb.h:34:0,
                 from cairo-xcb.h:46,
                 from cairo-xcb-private.h:42,
                 from cairo-xcb-connection-render.c:34:
C:/msys64/mingw64/include/xcb/xcb_windefs.h:36:10: fatal error: X11/XTrans/Xtrans.h: No such file or directory
 #include <X11/XTrans/Xtrans.h>
          ^~~~~~~~~~~~~~~~~~~~~
compilation terminated.
In file included from C:/msys64/mingw64/include/xcb/xcb.h:34:0,
                 from cairo-xcb.h:46,
                 from cairo-xcb-private.h:42,
                 from cairo-xcb-surface-core.c:35:
C:/msys64/mingw64/include/xcb/xcb_windefs.h:36:10: fatal error: X11/XTrans/Xtrans.h: No such file or directory
 #include <X11/XTrans/Xtrans.h>
          ^~~~~~~~~~~~~~~~~~~~~
compilation terminated.
In file included from C:/msys64/mingw64/include/xcb/xcb.h:34:0,
                 from cairo-xcb.h:46,
                 from cairo-xcb-private.h:42,
                 from cairo-xcb-resources.c:57:
C:/msys64/mingw64/include/xcb/xcb_windefs.h:36:10: fatal error: X11/XTrans/Xtrans.h: No such file or directory
 #include <X11/XTrans/Xtrans.h>
          ^~~~~~~~~~~~~~~~~~~~~
compilation terminated.
In file included from C:/msys64/mingw64/include/xcb/xcb.h:34:0,
                 from cairo-xcb.h:46,
                 from cairo-xcb-private.h:42,
                 from cairo-xcb-surface-render.c:34:
C:/msys64/mingw64/include/xcb/xcb_windefs.h:36:10: fatal error: X11/XTrans/Xtrans.h: No such file or directory
 #include <X11/XTrans/Xtrans.h>
          ^~~~~~~~~~~~~~~~~~~~~
compilation terminated.
In file included from C:/msys64/mingw64/include/xcb/xcb.h:34:0,
                 from cairo-xcb.h:46,
                 from cairo-xcb-surface.c:43:
C:/msys64/mingw64/include/xcb/xcb_windefs.h:36:10: fatal error: X11/XTrans/Xtrans.h: No such file or directory
 #include <X11/XTrans/Xtrans.h>
          ^~~~~~~~~~~~~~~~~~~~~
compilation terminated.
make[3]: *** [Makefile:2348: cairo-xcb-screen.lo] Error 1
make[3]: *** Waiting for unfinished jobs....
In file included from C:/msys64/mingw64/include/xcb/xcb.h:34:0,
                 from cairo-xcb.h:46,
                 from cairo-xcb-private.h:42,
                 from cairo-xcb-connection.c:35:
C:/msys64/mingw64/include/xcb/xcb_windefs.h:36:10: fatal error: X11/XTrans/Xtrans.h: No such file or directory
 #include <X11/XTrans/Xtrans.h>
          ^~~~~~~~~~~~~~~~~~~~~
compilation terminated.
make[3]: *** [Makefile:2348: cairo-xcb-connection-core.lo] Error 1
make[3]: *** [Makefile:2348: cairo-xlib-xcb-surface.lo] Error 1
make[3]: *** [Makefile:2348: cairo-xcb-surface-core.lo] Error 1
make[3]: *** [Makefile:2348: cairo-xcb-connection-render.lo] Error 1
make[3]: *** [Makefile:2348: cairo-xcb-resources.lo] Error 1
make[3]: *** [Makefile:2348: cairo-xcb-surface-render.lo] Error 1
make[3]: *** [Makefile:2348: cairo-xcb-surface.lo] Error 1
make[3]: *** [Makefile:2348: cairo-xcb-connection.lo] Error 1
make[3]: Leaving directory '/home/NateJoe/ff-build/work/mingw64/cairo-1.15.6/src'
make[2]: *** [Makefile:1947: all] Error 2
make[2]: Leaving directory '/home/NateJoe/ff-build/work/mingw64/cairo-1.15.6/src'
make[1]: *** [Makefile:904: all-recursive] Error 1
make[1]: Leaving directory '/home/NateJoe/ff-build/work/mingw64/cairo-1.15.6'
make: *** [Makefile:755: all] Error 2
!!! Build failed at: cairo-1.15.6

NateJoe@TinkeringAround MINGW64 ~/ff-build
$

FYI, I had wiped out MSYS2 and recreated it before by build attempt wherein I had myself expanded the binary and source archives that one places in original-archives. I could start over again if you wish. I've kind of got that process down now. LOL.

**Comment by JaiSoone** [comment]
P.S. In attempting this rebuild, the  .pacman-mingw64-installed got recreated!

**Comment by jtanx** [comment]
Ahh I see... I think. Thanks for persisting through this :)

I haven't set up the package dependencies properly, so I think you'll have to install the xtrans package manually, like:

```
pacman -S mingw-w64-x86_64-xtrans
```

----
Edit:
While you're at it, might as well try installing all the other dependencies to see if there's any that I've missed:

```
pacman -S --needed mingw-w64-x86_64-{gcc,glib2,pkg-config,libpng,freetype,fontconfig,libx11,libxrender,libxext,libxcb,libxft}
```

It'd be great if you could report any packages that were installed as a result of that command.

----

That should hopefully resolve the compile errors you're seeing. I'll need to fix up the package dependencies a bit, I think.

 `.pacman-mingw64-installed` being recreated is expected - ffbuild.sh uses its presence to know if it can skip having to reinstall its dependencies to help speed up the process when running the script multiple times.

**Comment by JaiSoone** [comment]
Ok, I'll try it and let you know what gets newly loaded as a result.
Attached are a detailed set of notes as to what I'm exactly doing
as I try following the Build Procedure. Note my step 5b below.
I don't know if it is a problem or not.

My steps in creating the fontforge 64 bit build.

1) Down loaded the installer: msys2-x86_64-20161025.exe
2) Ran this installer
    a) installed into C:\msys64
    b) had it put shortcuts in Startup
    c) allowed it to run MSYS2 64bit now.
    d) all I see in this console window is:
	NateJoe@TinkeringAround MSYS ~
	$
3) exit this new console window
4) a) cut and pasted 3 short cuts from:
	C:\Users\NateJoe\AppData\Roaming\Microsoft\Windows\Start Menu\Programs\Startup\
	to:
	C:\Users\NateJoe\Desktop\
    b) the names of these 3 shortcuts were
         1) MSYS2 MSYS
         2) MSYS2 MinGW 32-bit
         3) MSYS2 MinGW 64-bit
    c) these are shortcuts respectively to:
         1) C:\msys64\msys2_shell.cmd -msys
         2) C:\msys64\msys2_shell.cmd -mingw32
         3) C:\msys64\msys2_shell.cmd -mingw64
    d) All 3 shortcuts start in: C:\msys64
    e) For the sake of brevity, I renamed these 3 shortcut respectively to:
         1) MSYS2
         2) MinGW 32-bit
        3) MinGW 64-bit
       via the name field of the General tab of the Properties window for each of these shortcuts.
5) Clicked on the MinGW 64-bit shortcut to create a command or console window to the MSYS2 environment.
    a) At this point, the only .bat file in C:\msys64\ is an autorebase.bat file.
    b) This means that I cannot do what you ask to be done in Step 1 Substep 3 of the Build Procedure.
    c) Also, in neither step 2d above or in this step am I ever asked to close the terminal window and
        open a new one to continue the initialization process as is mentioned in Step1 Substep 4 of the Build Procedure.
    d) I am asked such a question when I get to Step 3 of the Build Procedure, and I respond in a similar manner
         at that point, but the Build Procedure does not indicate that such an interaction will happen at this point!
6) Nonetheless, proceeding to Step 2 of the Build Procedure, I enter the following command into my command window:
     pacman -Syy
7) Corresponding to Step3 of the Build Procedure, I enter the command
    pacman -Suu
    a) Proceed with installation? [Y/N] Y
    b) (5/5) upgrading pacman                             [#####################] 100%
warning: terminate MSYS2 without returning to shell and check for updates again
warning: for example close your terminal window instead of calling exit
    c) I click on the Close button in the upper right corner of command window
    d) I get a dialog box 
           mintty
                     Processes are runnning is session: Close anyway?
           to which I click on OK
     e) I now see in the command window:
          Hangup signal recieved.
          but the window does not go away.
      f) I do see from the Windows Task Manager that the bash application has gone away.
      g) So, I click on the close button on the terminal window again.
      h) To which I get a popup window saying;
           Terminl

            Terminal is not responding
             If you close the program, you might lose information.

             So, I clicked "Close the program"
             and the window went away.
8) Clicked on the MinGW 64-bit shortcut to create another command or console window
	to the MSYS2 environment.
9) Rerun the command:
    pacman -Suu
    a) Proceed with installation? [Y/N] Y
    b) Completed successfully.
10) Corresponding to Step 4 of the Build Procedure, I entered the command
      a) pacman -S git
      b) Proceed with installation? [Y/N] y
      c) Completed successfully.
11) Corresponding to Step 5 of the Build Procedure, I entered the command
      a) git clone https://github.com/jtanx/fontforgebuilds ff-build
      b) Completed successfully
12) There is no /etc/.pacman init file of some sort. Only a pacman.conf file.
13) Corresponding tp Step 6 of the Build Procedure, I entered the commands:
     a) git checkout msys2-configs
     b) ./patch-msys2-config.sh
         Got the warning:
         Done. You may have to restart the
         terminal for the changes to work.

     c) git checkout master
     d) Everythings looks good.
     e) Entered the command
          exit
14)  Clicked on the MinGW 64-bit shortcut to create another command or console window
	to the MSYS2 environment.
15) Corresponding to Step 7 of the Build Procedure, I place the following archives in their proper places.
      a) original-archives/binaries/  1) potrace-1.11.win64.tar.gz
                                                       2) VcXsrv-1.17.0.0-x86_64-minimal.tar.xz
      b) original-archives/sources    1) cairo-1.15.6.tar.xz
                                                       2) pango-1.40.7.tar.xz
                                                       3) freetype-2.8.tar.gz
16) Modified ffbuild.sh: changed 1.40.6 reference to 1.40.7 (only one place to change).

17) In the command window:
       cd ff_build
18) ./ffbuild.sh -d
       Failed at about the same place in cairo!

**Comment by JaiSoone** [comment]
Here is what happened when I entered the commands requested:

NateJoe@TinkeringAround MINGW64 ~
$ pacman -S mingw-w64-x86_64-xtrans
resolving dependencies...
looking for conflicting packages...

Packages (1) mingw-w64-x86_64-xtrans-git-r252.2836667-1

Total Download Size:   0.04 MiB
Total Installed Size:  0.23 MiB

:: Proceed with installation? [Y/n]

:: Retrieving packages...
 mingw-w64-x86_64-xtrans-git-r252...    43.8 KiB   340K/s 00:00 [#################################] 100%
(1/1) checking keys in keyring                                  [#################################] 100%
(1/1) checking package integrity                                [#################################] 100%
(1/1) loading package files                                     [#################################] 100%
(1/1) checking for file conflicts                               [#################################] 100%
(1/1) checking available disk space                             [#################################] 100%
:: Processing package changes...
(1/1) installing mingw-w64-x86_64-xtrans-git                    [#################################] 100%

That seemed to work!
Now for the big command to see what also might be missing.


$ pacman -S --needed mingw-w64-x86_64-{gcc,glib2,pkg-config,libpng,freetype,fontconfig,libx11,libxrender,libxext,libxcb,libxft}
warning: mingw-w64-x86_64-gcc-7.1.0-2 is up to date -- skipping
warning: mingw-w64-x86_64-glib2-2.52.3-1 is up to date -- skipping
warning: mingw-w64-x86_64-libpng-1.6.30-1 is up to date -- skipping
warning: mingw-w64-x86_64-freetype-2.8-1 is up to date -- skipping
warning: mingw-w64-x86_64-fontconfig-2.12.4-1 is up to date -- skipping
warning: mingw-w64-x86_64-libx11-git-r1325.7d2010fe-1 is up to date -- skipping
warning: mingw-w64-x86_64-libxrender-git-r82.f4f32f1-1 is up to date -- skipping
warning: mingw-w64-x86_64-libxext-git-r205.a07b4bb-1 is up to date -- skipping
warning: mingw-w64-x86_64-libxcb-git-r663.d10194a-1 is up to date -- skipping
warning: mingw-w64-x86_64-libxft-git-r102.c418dc7-1 is up to date -- skipping
resolving dependencies...
looking for conflicting packages...

Packages (1) mingw-w64-x86_64-pkg-config-0.29.2-1

Total Download Size:   0.23 MiB
Total Installed Size:  1.31 MiB

:: Proceed with installation? [Y/n] y
:: Retrieving packages...
 mingw-w64-x86_64-pkg-config-0.29.2-1-any              238.9 KiB  1201K/s 00:00 [#############################################] 100%
(1/1) checking keys in keyring                                                  [#############################################] 100%
(1/1) checking package integrity                                                [#############################################] 100%
(1/1) loading package files                                                     [#############################################] 100%
(1/1) checking for file conflicts                                               [#############################################] 100%
(1/1) checking available disk space                                             [#############################################] 100%
:: Processing package changes...
(1/1) installing mingw-w64-x86_64-pkg-config                                    [#############################################] 100%

NateJoe@TinkeringAround MINGW64 ~

**Comment by JaiSoone** [comment]
So, I restarted the build and go somewhat further and ran into:
Building VcXsrv_util...
Building run_fontforge...
Extracting the FreeType 2.8 source...
tar: /home/NateJoe/ff-build/original-archives/sources//freetype-2.8.tar.bz2: Cannot open: No such file or directory
tar: Error is not recoverable: exiting now
!!! Build failed at: FreeType2 extraction

Oops, I put the .gz archive in there! My bad.
Got the right one in place and crancked it up again. Ran, and ran, and ran.
I'm impressed. Used up to 6 Gb of memory; ran my CPU utilization to 100% at around 2.65 Mhz.
Got all 4 cups and I suppose all 8 threads cranking! And no disc I/O at all!

Ran in to the next problem though, which is:

Installing VcXsrv_util...
Installing run_fontforge...
Copying UI fonts...
No UI fonts specified, copying some standard ones...
Copying the custom resource file...
Copying sfd icon...
Copying the Python executable and libraries...
Python folder not found - running 'strip-python'...
cp: cannot stat '/MINGW64/lib/python3.5': No such file or directory
cp: cannot stat '/home/NateJoe/ff-build/original-archives/binaries//python3.5': No such file or directory
!!! Build failed at: Python folder could not be copied

So, I looked in '/MINGW64/lib/ and see both a python2.7 directory and a python3.6 directory.
The python2.7 directory doesn't have a lot of stuff in it, but the python3.6 directory sure does.
Should that be a reference to the 3.6 version? If so, where should the change be made?

**Comment by JaiSoone** [comment]
So, I changed the 3.5 python reference in ffbuild.sh for 64 bit builds to 3.6. I then found out that I had to
modify strip-python.sh also. Now I'm stopping at:

Python folder not found - running 'strip-python'...
Stripping Python cache files (*.pyc,*.pyo,__pycache__)...
Copying the Python extension dlls...
Generating the version file...
Adding in debugging utilities...
cp: cannot stat '/home/NateJoe/ff-build/original-archives/binaries//gdb-64.exe': No such file or directory
!!! Build failed at: Couldn't copy GDB

I guess this is happening because I'm compiling with the -d switch. Almost there. I'll leave this until
tomorrow.

**Comment by jtanx** [comment]
Good stuff! Yeah, unfortunately the 64-bit build doesn't get as much love/attention as the 32-bit build - your changes to use 3.6 sound good.

The `-d` switch is even less used (practically only once when it was first written) - to solve that error, download the 64bit version of GDB from here: http://www.equation.com/servlet/equation.cmd?fa=gdb, rename it to `gdb-64.exe`, and place it in `/home/NateJoe/ff-build/original-archives/binaries/`.

**Comment by JaiSoone** [comment]
More progress. I'm now failing on:

Copying the Python executable and libraries...
Skipping python library copy because folder already exists, and copying is slow.
Stripping Python cache files (*.pyc,*.pyo,__pycache__)...
Copying the Python extension dlls...
Generating the version file...
Adding in debugging utilities...
cp: cannot stat '/home/NateJoe/ff-build/original-archives/binaries//wtee.exe': No such file or directory
!!! Build failed at: Couldn't copy wtee

**Comment by jtanx** [comment]
I'm a bit hazy on where I got that one from, probably here: https://code.google.com/archive/p/wintee/downloads

**Comment by JaiSoone** [comment]
Success! I built it through to completion. I haven't tried it yet though.
Question: are there any optional features that have compile time switches to turn them on/off?
If so, are they documented anywhere?
I think I found what I'm looking for in:
~/ff-build/work/mingw64/fontforge/configure --help
The question now becomes, How do you pass such switches through ./ffbuild.sh to the compilation process? Also, is Unicode 10.0 integrated herein already, or is that an update which I could do my self?
I see in other places how one might go about doing this. Just curious.

**Comment by jtanx** [comment]
Good job getting it working!

>  How do you pass such switches through ./ffbuild.sh to the compilation process? 

Generally, the compile time configure flags to fontforge are hardcoded by ffbuild.sh around here: https://github.com/jtanx/fontforgebuilds/blob/master/ffbuild.sh#L629

However, now that you've completed a build. you can just cd directly into `~/ff-build/work/mingw64/fontforge`, modify it as you wish (e.g. update it to Unicode 10.0, which I think isn't integrated, just going off the git log), then rerun `ffbuild.sh`. You can actually run `make` and `make install` directly in that folder yourself, but for the `./configure` step and to bundle all the files into `ReleasePackage`, you need to run `./ffbuild.sh`. Technically, you can also edit the autoconf/automake files (e.g. configure.ac, *.m4, Makefile.am), and as long as you don't call `./configure` directly, the build system should pick up those changes, and everything should still just work.

If any modifications are required at the `./configure` stage, then these need to be made directly to `ffbuild.sh`, and from there you can run `./ffbuild.sh --reconfigure`, which will rerun the configure step before compiling anything. 

> Also, is Unicode 10.0 integrated herein already, or is that an update which I could do my self?

I'm not sure if this step was ever made to work on Windows (I haven't tried this out myself; @JoesCat normally does it), but assuming [the instructions](https://github.com/fontforge/fontforge/wiki/How-To-Update-Unicode-Support) are still correct, it would be something like:

```bash
cd ~/ff-build/work/mingw64/fontforge
cd Unicode
make makeutype
wget http://unicode.org/Public/UNIDATA/LineBreak.txt;
wget http://unicode.org/Public/UNIDATA/NamesList.txt;
wget http://unicode.org/Public/UNIDATA/PropList.txt;
wget http://unicode.org/Public/UNIDATA/UnicodeData.txt;
./makeutype
cd ~/ff-build
./ffbuild.sh
```

Finally, just note that `ffbuild.sh` set `~/ff-build/work/mingw64/fontforge` to track my fork - https://github.com/jtanx/fontforge, instead of https://github.com/fontforge/fontforge, just because there used to be some Windows-specific patches that needed to be applied. You can change it to use the official repo instead by something like:

```bash
cd ~/ff-build/work/mingw64/fontforge
git remote remove origin
git remote add origin https://github.com/fontforge/fontforge
git pull origin master                                 # (assuming you haven't made any modifications to FontForge)
```

**Comment by JaiSoone** [comment]
FYI: I tried making the makeutype  command file, but I'm missing a .h file:

```
NateJoe@TinkeringAround MINGW64 ~/ff-build/work/mingw64/fontforge/Unicode
$ make makeutype
x86_64-w64-mingw32-gcc -DWIN32 -I/home/NateJoe/ff-build/target/mingw64//include -I/mingw64/include -I/usr/local/include -I/include -g -Wall -DWIN32 -I/home/NateJoe/ff-build/target/mingw64//include -I/mingw64/include -I/usr/local/include -I/include -g -Wall -I.. -I.. -L/home/NateJoe/ff-build/target/mingw64//lib -L/mingw64/lib -L/usr/local/lib -L/lib -lws2_32  makeutype.c   -o makeutype
makeutype.c:53:10: fatal error: fontforge-config.h: No such file or directory
 #include <fontforge-config.h>
          ^~~~~~~~~~~~~~~~~~~~
compilation terminated.
make: *** [<builtin>: makeutype] Error 1
```

**Comment by JoesCat** [comment]
On July 27, 2017 08:09:15 PM JaiSoone wrote:
> FYI: I tried making the makeutype  command file, but I'm missing a .h
> file:

Hi JaiSoone, at the moment my fontforge build is parked off-line and a 
little out of date compared to the recent builds, plus I've still got 4 big 
tasks to get out of the way first - so no time to look at fontforge ATM, 
....but from memory....(1) I believe you should make fontforge first.
./bootstrap; ./configure; make;
(to build fontforge-config.h)

(2) Then you need to go into directory Unicode\
and then next, (3) you need to build ./makeutype before you can use it:
Run "gcc -s -I../inc -o makeutype makeutype.c"
if fontforge-config.h is in another directory, just add another -I../here



> NateJoe@TinkeringAround MINGW64
> ~/ff-build/work/mingw64/fontforge/Unicode $ make makeutype
> x86_64-w64-mingw32-gcc -DWIN32
> -I/home/NateJoe/ff-build/target/mingw64//include -I/mingw64/include
> -I/usr/local/include -I/include -g -Wall -DWIN32
> -I/home/NateJoe/ff-build/target/mingw64//include -I/mingw64/include
> -I/usr/local/include -I/include -g -Wall -I.. -I..
> -L/home/NateJoe/ff-build/target/mingw64//lib -L/mingw64/lib
> -L/usr/local/lib -L/lib -lws2_32  makeutype.c   -o makeutype
> makeutype.c:53:10: fatal error: fontforge-config.h: No such file or
> directory #include <fontforge-config.h>
>           ^~~~~~~~~~~~~~~~~~~~
> compilation terminated.
> make: *** [<builtin>: makeutype] Error 1

**Comment by JoesCat** [comment]
On July 27, 2017 02:27:10 PM Jeremy Tan wrote:
>> Also, is Unicode 10.0 integrated herein already, or is that an update
>> which I could do my self?
> 
> I'm not sure if this step was ever made to work on Windows (I haven't
> tried this out myself; @JoesCat normally does it), but assuming [the
> instructions](https://github.com/fontforge/fontforge/wiki/How-To-Update
> -Unicode-Support) are still correct, it would be something like:

Something like that ;-)
   ...and some tweaking along the way.

Right now, it's not a smooth process, but you can compare what you get with 
what you see in 9.0.0. One of the things that need improvement is 
automating this process, so you're sort of having to run makeutype, run 
makebuildtables, run dump, and charset/generate.sh

charset/generate.sh is supposed to replace makebuildtables, but I found it 
misses a couple files, so run makebuildtables to supplement the shortages, 
and you may need to look at various files by eyeball to avoid conflicts.

Then there are also the copy-n-pastes that need to be done from Unicode to 
fontforge and fontforgeexe directories. These are the tricky ones since no 
automation exists yet, or due to new patching without realizing they were 
automated, we ended-up with files that can't simply be copy-n-pasted.

I suggest looking at any patch that has titles like unicode 9, unicode 8, 
unicode 7, etc, to help find anything easily missed (yes, I missed some, and 
yes, unfortunately there are probably some not listed that I still missed, 
for example, I believe fontforge/*ttf*.c have some tables based on 
unicode).

If you are finding yourself pulling your hair out, welcome to the club :-)
...but as bad as I make it sound now, I still recommend you give it a try 
to see what needs work and improvement.

Start with the 9.0 patch to get some idea of what you need to run, and 
supplement it with unicode 8,7,....patches to catch the stuff that manged to 
get missed. You will want to run makeutype, makebuildtables, dump, 
charset/generate.c (watch for conflicts with this last program, and compare 
old with new files for "everything". If something doesn't look right, you 
may need to backsearch 20140101 or 20120731.

Last but not least, a bit of patience, these particular Unicode programs 
need some attention to bring everything up to good running order.... but 
it's a task for another day (for now).

**Comment by jtanx** [comment]
> $ make makeutype

I can reproduce this error - it's actually not Windows specific. The makefile is just broken for that target (the target for the binary is commented out/not set up properly) - running `make makeutype` is actually trying to invoke the target for building `makeutype.c` directly, which doesn't work because the include paths aren't set up properly.

You'd have to change `Makefile.am` to include something like:

```make
bin_PROGRAMS = makeutype

makeutype_SOURCES = makeutype.c combiners.h
makeutype_CFLAGS = "-I$(top_builddir)/inc"
```

Then just run `make`. This is a bit of a hack though because normally you don't want to build `makeutype`; you only want to make it when modifying this unicode stuff.

**Comment by JaiSoone** [comment]
Well, a lot of food for thought. I can see that I'll have to read up on how this compilation process works. I think at the top of it is automake; so, I'll start there. Can't very well use it for a higher purpose until you an at least drive it around the block so to speak. QED on this thread though. Thanks so much.
